python/amici/ode_export.py

```diff
@@ -1036,7 +1036,18 @@ def __init__(self, verbose: Optional[Union[bool, int]] = False,
             }
 
         self._lock_total_derivative: List[str] = list()
-        self._simplify: Callable = simplify
+        self._simplify: Callable = None
+        if simplify is not None:
+            def cached_simplify(
+                expr: sp.Expr,
+                _simplified: Dict[str, sp.Expr] = {},
+                _simplify: Callable = simplify,
+            ) -> sp.Expr:
+                expr_str = str(expr)
+                if expr_str not in _simplified:
+                    _simplified[expr_str] = _simplify(expr)
```

Comment:
```suggestion
                    _simplified[id(expr)] = _simplify(expr)
```

Will that work? Would save the time for stringification and some memory...

Reply:
Maybe not. Will fail if a memory address happens to be reused for a different expression. Not sure how often that will happen, but it's potentially unsafe.